### DISCUSSION TITLE
Please support Korean fonts

### DISCUSSION THREAD

**Discussion by magicianaos** [dfint/.github#1] [Q&A] [answered]
DF is a very popular game in Korea right now.
The recently released Steam version of DF's Korean translation participation rate is very good.

However, the DF font in the Steam version does not support Korean.
Can you support the tool that applies to the steam version DF and the Korean language to be output?

**Comment by insolor** [comment]
Hi
Technically this is possible, but this needs at least bringing back of rendering text using TTF fonts, and this is very tricky task.
Probably I'll try to do this, but not in the nearest future.

**Reply by magicianaos** [reply]
Thanks for your answer :)
We can support you.
Please tell me if you need anything.

**Comment by trotsky1997** [comment]
Tarn is trying on SDL TrueType font rendering on steam edition.

**Reply by insolor** [reply]
From where do you know it?

**Reply by magicianaos** [reply]
![20221213_101621](https://user-images.githubusercontent.com/6631032/207203452-ea5c3211-b090-4c8c-9244-aec189b7b701.png)
Game Main menu -> About DF

I don't know how to make the font like the picture below to add another language.
I need help.
![curses_640x300](https://user-images.githubusercontent.com/6631032/207203840-519425d2-3d7d-4347-ad9f-5fa8f913a10e.png)

**Reply by insolor** [reply]
This is just an outdated information. There is no more SDL_ttf.dll, no ttf font, no true-type font options (TrueType Font = TTF). Probably at some point of development of the steam version is was supported, but not anymore.

As for graphic font, anyone can edit the second half of the font (starting from the "funny" C letter, down below from it and to the right) to add their regional symbols, or an entire alphabet (if it is compact). The second half can contain 128 symbols, and in a text these symbols will be shown from left to right, so no compound symbols. So I presume you cannot use such font for Korean.

**Reply by trotsky1997** [reply]
> From where do you know it?

Sorry,I misunderstanded the update log, there was no sdl_ttf.dll now for steam edition

**Reply by trotsky1997** [reply]
> This is just an outdated information. There is no more SDL_ttf.dll, no ttf font, no true-type font options (TrueType Font = TTF). Probably at some point of development of the steam version is was supported, but not anymore.
> 
> As for graphic font, anyone can edit the second half of the font (starting from the "funny" C letter, down below from it and to the right) to add their regional symbols, or an entire alphabet (if it is compact). The second half can contain 128 symbols, and in a text these symbols will be shown from left to right, so no compound symbols. So I presume you cannot use such font for Korean.

By the way, can you grant me a coordinator permission on the transifex platform? I'm working on chinese translation with friends and we may need a coordinator for the management of chinese localizations. thanks

**Reply by trotsky1997** [reply]
> As for graphic font, anyone can edit the second half of the font (starting from the "funny" C letter, down below from it and to the right) to add their regional symbols, or an entire alphabet (if it is compact). The second half can contain 128 symbols, and in a text these symbols will be shown from left to right, so no compound symbols. So I presume you cannot use such font for Korean.

 As for the compound symbols, there were some potential similar solutions for English software with only single-byte support like Crusader kings, EU4, older GTA series.
The main difficulty of rendering double-byte compound symbols like CJK fonts and arabics is the rendering function of program only support the addressing of the single-byte symbols. What we need to do is hacking the addressing process and rendering function to add double-byte support.
This may lead to a violation of software's liscense policy. We can have a further talk if you have interest on.

**Reply by trotsky1997** [reply]
> ![20221213_101621](https://user-images.githubusercontent.com/6631032/207203452-ea5c3211-b090-4c8c-9244-aec189b7b701.png) 游戏主菜单->关于DF
> 
> 我不知道如何制作如下图所示的字体来添加另一种语言。 我需要帮助。 ![诅咒_640x300](https://user-images.githubusercontent.com/6631032/207203840-519425d2-3d7d-4347-ad9f-5fa8f913a10e.png)

Sorry my Asian fellow, DF only support single-byte font's rendering besides CJKs. We may need some hack to add double-byte support.

**Reply by insolor** [reply]
> By the way, can you grant me a coordinator permission on the transifex platform? I'm working on chinese translation with friends and we may need a coordinator for the management of chinese localizations. thanks

I rarely make anyone a coordinator. I do it only when I trust that person.

I can give you reviewer's permissions, but you need to tell me in translation to which language (or languages) you are intrested and your profile name on transifex.

**Comment by trotsky1997** [comment]
ooo，i have analysed the coding page of DF steam edition, its encoding type was UTF-8,and in fact, the bitmap graphic image is enough to store the Chinese chracters in 8x12px!
I have been wrong in above comment for that the font rendering function of DF steam edition already support addressing of double-byte chracter. And what we need to do is add new chracter to the the bitmap graphic font image!
binggo! I'm trying.

**Comment by trotsky1997** [comment]
![无标题](https://user-images.githubusercontent.com/30512160/207700217-8630ddd6-4842-4cff-8c24-677160afcee0.png)
That was a demo for Chinese character's rendering with 8x12px,it would be better in higher pixel

**Reply by Juijote** [reply]
You can find some free pixel fonts here, I tested a couple of them, and they were pretty good .

https://www.maoken.com/?s=%E5%83%8F%E7%B4%A0

![s](https://user-images.githubusercontent.com/22154581/207999458-4e8bd183-d81e-4e4b-aabb-73ab00bd738b.png)

![s (1)](https://user-images.githubusercontent.com/22154581/207999472-4eab3316-d170-441b-84ce-f34648597a66.png)

![s (2)](https://user-images.githubusercontent.com/22154581/207999483-8c26d7ba-89d2-4ee3-b8f3-bfdee7511bf6.png)

**Reply by p2mm2p** [reply]
The display effect is quite satisfactory.

**Comment by trotsky1997** [comment]
![image](https://user-images.githubusercontent.com/30512160/207703608-a8372e17-084a-4d6d-82b8-17250a0b5127.png)

**Comment by raydan** [comment]
由於只有 8x12px, 用圖片是沒有辦法顯示一個清晰的中文字

以下是個人測試後的結果
把 SDL ttf 用在 steam 上是可以顯示中文字

![df_translate](https://user-images.githubusercontent.com/5609845/209445373-90d78d5a-b330-414b-b3ad-e337c5fdae3a.png)

**Reply by insolor** [reply]
@raydan have you managed to use sdl_ttf somehow?

**Reply by raydan** [reply]
i use df-translate-hook for reference
load SDL ttf & replace the texrure

**Reply by trotsky1997** [reply]
> i use df-translate-hook for reference load SDL ttf & replace the texrure

期待进展

**Reply by raydan** [reply]
now i can use the transifex po file to translate & display replaced word in game

"Set tree chopping orders."
the problem is the above string cut to two string "Set tree chopping " & "orders." pass to "addst"
which it unable to replace now, can not find any way to do that

i think it hard to display a clear chinese text right now even using SDL TTF.

![image1](https://user-images.githubusercontent.com/5609845/209466136-884ce48b-61b3-46f6-a830-94f4def14cf0.png)

![image2](https://user-images.githubusercontent.com/5609845/209466137-5834acb8-870e-436e-a6b9-7ec5d26d63f6.png)

korean text
![image3](https://user-images.githubusercontent.com/5609845/209466997-8c917880-72e1-444d-9406-b54736fd3974.png)

**Reply by insolor** [reply]
@raydan can you share your code somewhere, please?

**Reply by trotsky1997** [reply]
It's perfect now. maybe you can share your code on github

**Comment by raydan** [comment]
***ONLY FOR TESTING & POC, THERE ARE MEMORY LEAK***

compile, rename dll to "df-steam-translate-hook.dll" and place to Dwarf Fortress.exe folder
copy "file" folder content to Dwarf Fortress.exe folder
rename "df_chs.csv" to "df_lang.csv"
rename "font_chs.ttf" to "font.ttf"

And i used df libgraphics for code reference
source code tested in v50.04

[df_dll_test.zip](https://github.com/dfint/.github/files/10302547/df_dll_test.zip)

**Reply by magicianaos** [reply]
I can't compile.
Can you share compiled dll file?

**Reply by raydan** [reply]
should be able to compile without error, all files are there

P.S.
just find some bug in load_multi_pdim_2, modify disp_x, disp_y should be after the original function

**Reply by magicianaos** [reply]
![20221226_194253](https://user-images.githubusercontent.com/6631032/209539787-023e84a1-3af9-43d4-a39c-3efa5a308fc2.png)

 1. Compile dll_test.sln.
2. After rename the file name, copy the file to the 'DF' folder.
3. Run

But Korean is not output.

**Reply by raydan** [reply]
you have to use df-steam-hook-launcher.exe from
https://github.com/dfint/misc_archives/tree/main/df_steam_patcher

**Reply by magicianaos** [reply]
Thanks:)

**Reply by trotsky1997** [reply]
> Thanks:)
I built a binary for users without CPP compile envoirment
Just drop into DF root directory

[社区汉化补丁尝鲜.zip](https://github.com/dfint/.github/files/10303270/default.zip)

**Reply by shevernitskiy** [reply]
@raydan i trying to reproduce your work in [df-steam-hook](https://github.com/dfint/df-steam-hook) but getting crash during loading/creation new world. Main menu seems to works, but there are some artifacts during scrolling in about section (some parts of letters not being erased).
I already have texture generation with caching. But when I use `add_texture` hook with `my_screen` texture swapping i'm have increased cpu usage (~x3, 10% in total, 2-3% without it).
I'm trying to figure out what i miss. Maybe you could see my code in repo and give me a hint? 

UPD: it seems like increased cpu usage was caused by wrong texture sizes, so they overlaped each other..
and crash was caused by cache after game start loading mod suqence (new game/load game initializing)... seems like game clear all textures in this moment, but cache still gave reference to cleared one texture...
but one moment still bother me - add_texture hook... i thinks it's leaking... and we can't use it just like we did now...

UPD2: so, i just cached returned id from add_texture - seems to work;)

**Reply by raydan** [reply]
you may need to hook on textures::delete_texture to delete your cached texture
and hook graphicst::addcoloredst to make it like "addst" string replace, but i never found that address

**Reply by shevernitskiy** [reply]
> you may need to hook on textures::delete_texture to delete your cached texture and hook graphicst::addcoloredst to make it like "addst" string replace, but i never found that address

I made texture swapping only in `addchar`, and in other hooks (like addst) just switch flag to swap.
And since i caching texture_id from `add_texture`, i obsreving no leaks in this moment... maybe there are some undeleted textures, but not so many, cause it not so many chars is use - it's not critical at all..
Now I faced with problem, where i have to find address for hook to very begin of loading save/new game. When we go from main menu (chose save for load or create new world) to this loading screen, its seems like game perform global mem clearing. And then my cache contains references to unexisting objects. So I need to reset cache exactly in this moment as well as game mem clear.

UPD: so, i hooked into loops of interface rendering (main menu and all kinds of loading/creating game) for game state tracking... reseting my cache whenever state changed to "game" from "menu"...

**Reply by shevernitskiy** [reply]
@raydan I found addcoloredst - 0x784890...
Now I working on strange issue. There are screen and screen_top (for dialog windows). Also there are addst_top, addchar_top and screen_to_texid_top - whole pack for different tile matrix. I successfuly hooked in the "top" pack. But when I trying to open ingame statistics table (Nobles and administrator for example) and popup tutorial windows appears in same time, text of main screen level become black. Don't now why, maybe i erased color or tex?

UDP:  we need to use separate screen matrix for top layer, it solves an issue (or maybe we should find proper offset in main matrix for start point top layer)... but if we use just one screen matrix, layers may interfere - that corrupts data...

**Reply by mocksonline** [reply]
in 50.05 it doesnt work

**Reply by shevernitskiy** [reply]
> in 50.05 it doesnt work

offsets slighly different in 50.05

**Comment by insolor** [comment, marked as answer]
As of 2025: due to the technical complexity of the task, we are not going to add support for Chinese/Japanese/Korean. We are going to focus only on languages with compact alphabets with left-to-right writing direction, with not too complex letters.
You can try to use [dfint-rust-cjk](https://gitee.com/vizv/dfint-rust-cjk), but it is a separate project, and we cannot guarantee anything.
Upd: Try to use [DFI18n](https://steamcommunity.com/sharedfiles/filedetails/?id=3613958631) mod. Source code: https://github.com/DFI18n/dfi18n. Theoretically, it should support Korean, though I myself didn't test it.
